### ISSUE TITLE
Installing different dev version

### ISSUE THREAD

**Issue by abhatthal** [koekeishiya/yabai#1131]
I'm running Monterey 12.1 and using yabai v4 built from HEAD.
I updated to the latest HEAD (297156c02854f21a811f5593504226a040a60414) by reinstalling and it causes SimpleBar, an Ubersicht widget to work incorrectly. I want to go back to my previous version but I'm not sure how to downgrade my local HEAD or otherwise build from a specific commit.

Sorry if this is a stupid question. I looked all over for a solution and tried installing from an older `yabai.rb` but the program just segfaulted. I tried using the v3.3.10 bottle, but it continuously fails to inject and shuffles my apps.

Here's a screenshot. You can see how it shows apps from different workspaces when Safari is the only one there.
<img width="1929" alt="image" src="https://user-images.githubusercontent.com/41220185/150276762-4ba87529-bcfc-4b9b-a97a-098aebdec9e6.png">

**Comment by abhatthal** [comment]
nvm I figured out how to do it!

In case anyone else is in this situation:
$  brew install koekeishiya/formulae/yabai --HEAD
$ cd ~/Library/Caches/Homebrew/yabai--git
$ git checkout f403e609e32b4100494c5afb089d0010e7e4ef91
$ make -j1 install
$ brew unlink yabai
$ ln -s ~/Library/Caches/Homebrew/yabai--git/bin/yabai /usr/local/bin/yabai

Now running `yabai` my SimpleBar works correctly. I tried starting the service with brew, but it doesn't work.
The plist being used for the service is referring to a different binary, we can fix that.
$ cd /usr/local/opt/yabai/
$ cd bin && mv yabai yabai-old
$ ln -s ~/Library/Caches/Homebrew/yabai--git/bin/yabai /usr/local/opt/yabai/bin/yabai
$ brew services start koekeishiya/formulae/yabai


It's working without any issues. Figured it out on my own, so please let me know if this solution is a terrible idea.

**Comment by koekeishiya** [comment]
The current master runs fine for me. I don't think the problem you are having was related to it.

**Comment by koekeishiya** [comment]
v4 of yabai (and thus the master branch) is a breaking release; there are changes to the query output as mentioned here: https://github.com/koekeishiya/yabai/issues/775 but the latest version of simple-bar claim to support it.

**Comment by abhatthal** [comment]
> v4 of yabai (and thus the master branch) is a breaking release; there are changes to the query output as mentioned here: #775 but the latest version of simple-bar claim to support it.

I just updated to the latest version of simple-bar. For some reason I forgot I had to do this manually. Anyhow, with v4 I still have this issue, I'll open an issue with simple-bar.

**Comment by Jean-Tinland** [comment]
Hi!

`yabai -m query --windows` on 12.1 with yabai v4 (on the latest commit) seems to correctly return all windows but with an incorrect space index. All windows are always identified as being on the current workspace.

**Comment by FelixKratz** [comment]
https://github.com/koekeishiya/yabai/blob/297156c02854f21a811f5593504226a040a60414/src/window.c#L89
and 5cb55e398ede077d08344f25b9c71f8ccc0b2df2 lead to this.

**Comment by koekeishiya** [comment]
Right.. thanks for catching that

**Comment by koekeishiya** [comment]
Should be fixed again on the latest master.

**Comment by Jean-Tinland** [comment]
Yes! Thank you for this quick fix!

**Comment by zt1983811** [comment]
> nvm I figured out how to do it!
> 
> In case anyone else is in this situation:
> 
> ```
> $ brew install koekeishiya/formulae/yabai --HEAD
> $ cd ~/Library/Caches/Homebrew/yabai--git
> $ git checkout f403e609e32b4100494c5afb089d0010e7e4ef91
> $ make -j1 install
> $ brew unlink yabai
> $ ln -s ~/Library/Caches/Homebrew/yabai--git/bin/yabai /usr/local/bin/yabai
> ```
> 
> Now running `yabai` my SimpleBar works correctly. I tried starting the service with brew, but it doesn't work. The plist being used for the service is referring to a different binary, we can fix that.
> 
> ```
> $ cd /usr/local/opt/yabai/
> $ cd bin && mv yabai yabai-old
> $ ln -s ~/Library/Caches/Homebrew/yabai--git/bin/yabai /usr/local/opt/yabai/bin/yabai
> $ brew services start koekeishiya/formulae/yabai
> ```
> 
> It's working without any issues. Figured it out on my own, so please let me know if this solution is a terrible idea.

Thank you do much you saved my day.....

**Comment by zt1983811** [comment]
> Should be fixed again on the latest master.

I don't think so I just tried it this morning it still having problem
```
2022-01-24 13:50:07.460 yabai[17600:284213] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[__NSCFNumber _daisyFrame]: unrecognized selector sent to instance 0x7ff82c143735'
*** First throw call stack:
(
        0   CoreFoundation                      0x00007ff8119fcf0b __exceptionPreprocess + 242
        1   libobjc.A.dylib                     0x00007ff81175db9d objc_exception_throw + 48
        2   CoreFoundation                      0x00007ff811a81adb -[NSObject(NSObject) __retain_OA] + 0
        3   CoreFoundation                      0x00007ff811963db9 ___forwarding___ + 1406
        4   CoreFoundation                      0x00007ff811a27fd8 __forwarding_prep_1___ + 120
        5   AppKit                              0x00007ff814a99e35 -[NSScreen safeAreaInsets] + 58
        6   yabai                               0x000000010f5045d0 workspace_display_has_notch + 480
        7   yabai                               0x000000010f51105d display_bounds_constrained + 141
        8   yabai                               0x000000010f513082 view_update + 98
        9   yabai                               0x000000010f513eff view_create + 159
        10  yabai                               0x000000010f51cad8 space_manager_init + 280
        11  yabai                               0x000000010f526878 main + 1256
        12  dyld                                0x0000000113fc94fe start + 462
)
libc++abi: terminating with uncaught exception of type NSException
```

**Comment by zt1983811** [comment]
@abhatthal 
Do you have the problem of switch focus after install that version? 
For all other functions it works only this one seems not working
I have disabled System Integrity Protection.
```
➜  ~ yabai -m window --focus west
could not locate a westward managed window.
```

**Comment by abhatthal** [comment]
@zt1983811 
I'm presently on 91a42e747197680330a4cd9c65f932a060b56e4e. Only issues I'm aware of is transparency not working.

I can change focus using `yabai -m window --focus west` or any other direction.

Your second command gives me the same error.
Ultimately this is the issue:
<img width="1251" alt="image" src="https://user-images.githubusercontent.com/41220185/150860345-ad3b7671-9fc0-4b88-8685-36f72170ba6b.png">
We can't find an instance where focused == 1, so we return null when searching for an index, and we can't subtract 1 from null which is why we got an error.

If that was working and suddenly wasn't, it's likely because there was an API change. v4 is a breaking release and is still being worked on. I see that there are windows with the `has-focus` attribute

Using `.\"has-focus\" == true` in place of `.focused == 1` seems to work.


TLDR; use this instead:
```
» yabai -m query --spaces --space \
  | jq -re ".index" \
  | xargs -I{} yabai -m query --windows --space {} \
  | jq -sre "add | map(select(.minimized != 1)) | sort_by(.display, .frame.y, .frame.x, .id) | reverse | nth(index(map(select(.\"has-focus\" == true))) - 1).id" \
  | xargs -I{} yabai -m window --focus {}
```

**Comment by zt1983811** [comment]
> @zt1983811 I'm presently on [91a42e7](https://github.com/koekeishiya/yabai/commit/91a42e747197680330a4cd9c65f932a060b56e4e). Only issues I'm aware of is transparency not working.
> 
> I can change focus using `yabai -m window --focus west` or any other direction.
> 
> Your second command gives me the same error. Ultimately this is the issue: <img alt="image" width="1251" src="https://user-images.githubusercontent.com/41220185/150860345-ad3b7671-9fc0-4b88-8685-36f72170ba6b.png"> We can't find an instance where focused == 1, so we return null when searching for an index, and we can't subtract 1 from null which is why we got an error.
> 
> If that was working and suddenly wasn't, it's likely because there was an API change. v4 is a breaking release and is still being worked on. I see that there are windows with the `has-focus` attribute
> 
> Using `.\"has-focus\" == true` in place of `.focused == 1` seems to work.
> 
> TLDR; use this instead:
> 
> ```
> » yabai -m query --spaces --space \
>   | jq -re ".index" \
>   | xargs -I{} yabai -m query --windows --space {} \
>   | jq -sre "add | map(select(.minimized != 1)) | sort_by(.display, .frame.y, .frame.x, .id) | reverse | nth(index(map(select(.\"has-focus\" == true))) - 1).id" \
>   | xargs -I{} yabai -m window --focus {}
> ```

Wow this is really cool @abhatthal . It works for me the one you wrote. 
I am using the same commits but for some reason `yabai -m window --focus west` is not working. Maybe some permission issue somewhere but the one you suggest is work, I can switch next and pre window. Thanks so much for that
